### PR TITLE
Feature: Multiple selection

### LINKED ISSUE #199: auro-menu: multiselect
### Is your feature request related to a problem? Please describe.

As a business I want to give users the ability to select multiple options from menu items
So that users can select multiple options in the UI.

Blueprint:
https://www.figma.com/file/yMshXAiKZrkLDdiX8N4GKQ/Multi-Select-Combo-Box?type=design&node-id=114-3745&mode=design&t=R007FIBn8d6HGOOY-11

### Describe the solution you'd like

A menu with checkboxes so users can select as many as they want.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Exit criteria

1. Auro-menu feature matches the blueprint.
2. Create refinement meeting team on next step to align on implementation of next step.
3. As an outcome of the refiement a new issue is created to implement this as either `<auro-multiselect>` or `<auro-combobox multiselect>`
4. When Menu is Multiselect, value is an Array containing all selected menu options.

### PR DESCRIPTION
**Pull request by fajar-apri-alaska**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Resolves:** AlaskaAirlines/auro-menu#199 

## Summary:

The code diff introduces a new `multiSelect` attribute to the `AuroMenu` component, allowing for multiple options to be selected. Here's a breakdown of the functional differences:

1. **New Attribute: `multiSelect`**
   - **Definition**: Added a `multiSelect` Boolean attribute to the class, allowing the menu to support multiple selections.
   - **Usage**: 
     ```js
     this.multiSelect = undefined;
     ```

2. **Handling `value` Property**
   - **Before**: The `value` property was a single string representing the selected value.
   - **After**: The `value` property can be either a string or an array of strings when `multiSelect` is true.
     ```js
     if (!this.multiSelect) {
       this.selectByValue(this.value);
     } else {
       const values = JSON.parse(this.value);
       if (this.value && Array.isArray(values)) {
         values.forEach((val) => {
           this.selectByValue(val);
         });
       }
     }
     ```

3. **Selection State Management**
   - **New Methods**: Added methods to handle multiple selections:
     - `toggleSelectionState`: Toggles the selection state of an option.
     - `handleDeselectState`: Handles deselection of an option, updating the `value` array.
     - Modified `handleLocalSelectState` to support multiple selections by updating the `value` array.
     ```js
     toggleSelectionState(option) {
       if (option.hasAttribute('selected')) {
         this.handleDeselectState(option);
       } else {
         this.handleLocalSelectState(option);
       }
     }
     handleDeselectState(option) {
       option.removeAttribute('selected');
       option.classList.remove('active');
       option.ariaSelected = false;

       if (this.multiSelect && Array.isArray(this.value)) {
         this.value = this.value.filter((val) => val !== option.value);
         if (this.value.length === 0) {
           this.value = undefined;
         }
       }

       if (this.multiSelect && Array.isArray(this.optionSelected)) {
         this.optionSelected = this.optionSelected.filter((val) => val !== option);
         if (this.optionSelected.length === 0) {
           this.optionSelected = undefined;
         }
       }

       this.index = this.items.indexOf(option);
     }
     ```

4. **Resetting Option States**
   - **Before**: `resetOptionsStates` was always called when selecting an item.
   - **After**: `resetOptionsStates` is only called when `multiSelect` is false.
     ```js
     if (!this.multiSelect) {
       this.resetOptionsStates();
     }
     ```

5. **Event Handling**
   - Modified event handling to toggle selection state instead of always setting it:
     ```js
     if (!this.multiSelect) {
       this.handleLocalSelectState(option);
     } else {
       this.toggleSelectionState(option);
     }
     ```

The changes allow the `AuroMenu` component to handle multiple selections by using an array for the `value` property and providing methods to manage the selection and deselection of multiple options.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [ ] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: docs/partials/api.md [hunk 1]
```diff
@@ -74,6 +74,24 @@ The `auro-menu` component supports the use of the `matchWord` attribute to highl
 
 </auro-accordion>
 
+### Multiple
+
+`auro-menu` can have `multiple` attribute, to select multiple `auro-menuoption` elements.
+If `multiple` is enabled, `value` attribute became array value.
```
Comment (blackfalcon) [inline docs/partials/api.md]:
```suggestion
The `auro-menu` supports a multi-select option. To use, place the `multiSelect` attribute on the `<auro-menu>` element tag. When applied, the `value` attribute will become an Array versus String value. 
```

File: src/auro-menu.js [hunk 2]
```diff
@@ -15,7 +15,8 @@ import "mark.js/dist/mark.min";
  * @attr {String} matchWord - Specifies the a string used to highlight matched string parts in options.
  * @attr {Boolean} disabled - When true, the entire menu and all options are disabled;
  * @attr {Boolean} noCheckmark - When true, selected option will not show the checkmark.
- * @attr {String} value - Value selected for the menu.
+ * @attr {Boolean} multiple - When true, selected option can be multiple.
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
 * @attr {Boolean} multiSelect - When true, the selected option can be multiple.
```

File: src/auro-menu.js [hunk 3]
```diff
@@ -15,7 +15,8 @@ import "mark.js/dist/mark.min";
...
+ * @attr {String} value - Value selected for the menu, can be String or Array if multiple is true.
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
 * @attr {String} value - Value selected for the menu, can be String or Array if `multiSelect` is true.
```

File: src/auro-menu.js [hunk 4]
```diff
@@ -41,6 +42,7 @@ export class AuroMenu extends LitElement {
     this.noCheckmark = false;
     this.ready = false;
     this.optionActive = undefined;
+    this.multiple = undefined;
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
    this.multiSelect = undefined;
```

File: src/auro-menu.js [hunk 5]
```diff
@@ -62,6 +64,7 @@ export class AuroMenu extends LitElement {
       optionActive:   { type: Object },
       matchWord:      { type: String },
       ready:          { type: Boolean },
+      multiple:       { type: Boolean },
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
      multiSelect:       { type: Boolean },
```

File: src/auro-menu.js [hunk 6]
```diff
@@ -100,10 +103,26 @@ export class AuroMenu extends LitElement {
       this.markOptions();
     }
 
+
     if (changedProperties.has('value')) {
-      this.selectByValue(this.value);
+      if (!this.multiple) {
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
      if (!this.multiSelect) {
```

File: src/auro-menu.js [hunk 7]
```diff
@@ -164,6 +183,50 @@ export class AuroMenu extends LitElement {
...
+   * @private
+   * @return {void}
+   */
+  handleDeselectState(option) {
+    option.removeAttribute('selected');
+    option.classList.remove('active');
+    option.ariaSelected = false;
+
+    // remove from this.value if there is any, set to undefined if became empty array
+    if (this.multiple && Array.isArray(this.value)) {
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
    if (this.multiSelect && Array.isArray(this.value)) {
```

File: src/auro-menu.js [hunk 8]
```diff
@@ -164,6 +183,50 @@ export class AuroMenu extends LitElement {
...
+      this.value = this.value.filter((val) => val !== option.value);
+      if (this.value.length === 0) {
+        this.value = undefined;
+      }
+    }
+
+    // do the same for selected option DOM
+    if (this.multiple && Array.isArray(this.optionSelected)) {
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
    if (this.multiSelect && Array.isArray(this.optionSelected)) {
```

File: src/auro-menu.js [hunk 9]
```diff
@@ -174,8 +237,29 @@ export class AuroMenu extends LitElement {
     option.classList.add('active');
     option.ariaSelected = true;
 
-    this.value = option.value;
-    this.optionSelected = option;
+    if (this.multiple) {
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
    if (this.multiSelect) {
```

File: src/auro-menu.js [hunk 10]
```diff
@@ -208,7 +292,9 @@ export class AuroMenu extends LitElement {
     }
 
     if (this.items[this.index] && !this.items[this.index].hasAttribute('disabled')) {
-      this.resetOptionsStates();
+      if (!this.multiple) {
```
Comment (blackfalcon) [inline src/auro-menu.js]:
```suggestion
      if (!this.multiSelect) {
```

File: test/auro-menu.test.js [hunk 11]
```diff
@@ -264,8 +264,68 @@ describe('auro-menu', () => {
...
+
+      index++;
+    }
+    await elementUpdated(menuEl);
+    
+    expect(menuEl.optionSelected.length).to.equal(6);
+  })
 });
 
+it('test select/deselect options in multiple', async () => {
```
Comment (blackfalcon) [inline test/auro-menu.test.js]:
```suggestion
it('test select/deselect options in multiSelect', async () => {
```

File: test/auro-menu.test.js [hunk 12]
```diff
@@ -264,8 +264,68 @@ describe('auro-menu', () => {
...
+  // deselect first option
+  menuEl.dispatchEvent(new KeyboardEvent('keydown', {
+    bubbles: true,
+    composed: true,
+    'key': 'Enter'
+  }));
+  expect(menuEl.children[0].hasAttribute('selected')).to.equal(false);
+})
+
+it('test changing values after render in multiple', async () => {
```
Comment (blackfalcon) [inline test/auro-menu.test.js]:
```suggestion
it('test changing values after render in multiSelect', async () => {
```

File: test/auro-menu.test.js [hunk 13]
```diff
@@ -359,3 +419,18 @@ async function emptyItemsFixture() {
     </div>
   `);
 }
+
+async function multipleMenuFixture() {
+  return await fixture(html`
+    <div>
+      <auro-menu multiple aria-label="test">
```
Comment (blackfalcon) [inline test/auro-menu.test.js]:
```suggestion
      <auro-menu multiSelect aria-label="test">
```

File: test/auro-menu.test.js [hunk 14]
```diff
@@ -359,3 +419,18 @@ async function emptyItemsFixture() {
...
+        <auro-menuoption value="option 1">option 1</auro-menuoption>
+        <auro-menuoption value="option 2">option 2</auro-menuoption>
+        <auro-menuoption value="option 3">option 3</auro-menuoption>
+        <auro-menuoption value="lorem ipsum lorem ipsum">lorem ipsum lorem ipsum</auro-menuoption>
+        <auro-menuoption value="departures">Departures</auro-menuoption>
+        <auro-menuoption value="arrivals">Arrivals</auro-menuoption>
+      </auro-menu>
+    </div>
+  `);
+}
```
Comment (blackfalcon) [inline test/auro-menu.test.js]: Please add an empty return.

**Review by blackfalcon — CHANGES_REQUESTED** [review]

**Comment by blackfalcon** [comment]
While this is functionally living up to expectations, we have a pretty big a11y/UX issue. Please see the attached video. Now that we have this option to multi-select more than one menu item, using the keyboard arrow keys to move up and down the list to unselect any options is almost impossible. 

See in the video that the DOM is triggering active state, but the UI is NOT reflecting this feedback making this almost unseable to assistive device use. 

@leeejune we will need some assistance with this. 

May-14-2024 12-43-19

**Comment by leeejune** [comment]
@blackfalcon I see a few suggestions to address this:
1. We use checkboxes for multi select menu items. I don't know how the behavior is inherited, but I see the "press space to select" behavior being inherited. That should solve this issue of selecting with up/down arrows.
2. We implement the "press enter to select" behavior like how it is shown here: https://semantic-ui.com/modules/dropdown.html

File: test/auro-menu.test.js (hunk 14 above)
Comment (fajar-apri-alaska) [reply test/auro-menu.test.js]: Sorry, what do you mean by empty return for this one? it is currently returns the fixture.

**Comment by blackfalcon** [comment]
> @blackfalcon I see a few suggestions to address this:
> 
> 1. We use checkboxes for multi select menu items. I don't know how the behavior is inherited, but I see the "press space to select" behavior being inherited. That should solve this issue of selecting with up/down arrows.
> 2. We implement the "press enter to select" behavior like how it is shown here: https://semantic-ui.com/modules/dropdown.html

Some of the recommendations you are suggesting will be implemented with the new multi-select combobox element where the selected options will appear in a list above the dropdown. But that's not the issue I am bringing up here. 

I am electing to move this to a new issue where we can discuss a keyboard interaction solution. 

https://github.com/AlaskaAirlines/auro-formkit/issues/149

**Review by blackfalcon — DISMISSED** [review]
I updated this commit and PR to fit an evolving specification the team is going for. Please let me know if you have any questions. 

Also, please look into this weird warning coming from the test logs on a build

https://github.com/AlaskaAirlines/auro-menu/actions/runs/9166453632/job/25201867531#step:6:25

**Comment by fajar-apri-alaska** [comment]
> I updated this commit and PR to fit an evolving specification the team is going for. Please let me know if you have any questions.
> 
> Also, please look into this weird warning coming from the test logs on a build
> 
> https://github.com/AlaskaAirlines/auro-menu/actions/runs/9166453632/job/25201867531#step:6:25

I am aware of this as well.
Since there is a [part of the code](https://github.com/AlaskaAirlines/auro-menu/blob/eef024c2a613343f6a95830bb7aca1910191f29c/src/auro-menu.js#L111-L120) where it reads the `value` attributes (which it can be an array). So I am putting that inside try catch function to avoid function break.

**Comment by jason-capsule42** [comment]
@sourcery-ai review

**Comment by chrisfalaska** [comment]
@blackfalcon I intend on picking this up and integrating it into `auro-formkit`.

**Comment by chrisfalaska** [comment]
This work has been moved to `auro-formkit` and will be completed in https://github.com/AlaskaAirlines/auro-formkit/issues/123

**Pull request closed without merging**